LFS-1085: Refactor query.jsx as a functional component; LFS-1151: Show no results when typing space in the vocabulary field

veronikaslc opened the pull request:
To test, check that there are no regressions in:
* feedback to the user when vocabulary is not yet installed and the user is typing in the suggested input
* adding a non-predefined answer to vocabulary questions with input
* getting the information about a term by clicking the info button
* browsing a vocabulary
* filtering by vocabulary questions in the Forms table

Also Fixes [LFS-1151](https://phenotips.atlassian.net/browse/LFS-1151) - Show no results when typing space in the vocabulary field

Commit:
LFS-1085: Refactor query.jsx as a functional component

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -516,91 +307,189 @@ class VocabularyQuery extends React.Component {
+          infoID={infoID}
+          infoDefinition={infoDefinition}
+          infoAlsoKnownAs={infoAlsoKnownAs}
+          infoTypeOf={infoTypeOf}
```
These could be passed as an object instead of individual props

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -516,91 +307,189 @@ class VocabularyQuery extends React.Component {
+          menuPopperRef={menuPopperRef}
+          infoVocabURL={infoVocabURL}
+          infoVocabDescription={infoVocabDescription}
+          infoVocabAcronym={infoVocabAcronym}
```
These could be passed as an object instead of individual props

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -516,91 +307,189 @@ class VocabularyQuery extends React.Component {
+        <InfoBox
+          termInfoVisible={termInfoVisible}
+          anchorEl={infoAnchor}
+          infoAboveBackground={infoAboveBackground}
```
This prop seems to depend on `nrowserOpeened`, can we move it inside `InfoBox`?

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -433,81 +232,73 @@ class VocabularyQuery extends React.Component {
-      || this.state.browserOpened) {
+  let closeAutocomplete = event => {
+    if ((menuPopperRef && menuPopperRef.current.contains(event.target))
+      || (infoRef && infoRef.current && infoRef.current.contains(event.target))
```
Refs are always defined, but the current value may not be, so a better check would be:

```suggestion
    if (menuPopperRef?.current?.contains(event.target)
      || infoRef?.current?.contains(event.target)
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -516,91 +307,189 @@ class VocabularyQuery extends React.Component {
-      || (this.infoRef && this.infoRef.contains(event.target))) {
+  let clickAwayInfo = (event) => {
+    if ((menuPopperRef && menuPopperRef.current.contains(event.target))
+      || (infoRef && infoRef.current && infoRef.current.contains(event.target))) {
```
```suggestion
    if (menuPopperRef?.current?.contains(event.target)
      || infoRef?.current?.contains(event.target)) {
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -516,91 +307,189 @@ class VocabularyQuery extends React.Component {
+  if (disabled) {
+    // Alter our text to either the override ("Please select at most X options")
+    // or empty it
+    anchorEl.currentl.value = disabled ? props.overrideText : "";
```
```suggestion
    anchorEl.current.value = disabled ? props.overrideText : "";
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/infoBox.jsx:
```diff
@@ -0,0 +1,140 @@
+  const { classes } = props;
+
+  let clickAwayInfo = (event) => {
+    if ((props.menuPopperRef && props.menuPopperRef.current?.contains(event.target))
+      || (props.infoRef && props.infoRef.current?.contains(event.target))) {
```
```suggestion
    if (props?.menuPopperRef.current?.contains(event.target)
      || props?.infoRef.current?.contains(event.target)) {
```

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Code optimisation

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Code optimisation

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Typo fix

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Code optimisation

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

Commit:
LFS-1085: Refactor query.jsx as a functional component
Addressed codereview comments

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -50,67 +51,67 @@ const MAX_RESULTS = 10;
+
+  const { classes, defaultValue, disabled, inputRef, noMargin, isNested, onChange, onInputFocus, placeholder, label, value } = props;
+
+  const [suggestions, setSuggestions] = useState([]);
```
```suggestion
```
Unused

veronikaslc replied:
Used for holding suggestions

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -50,67 +51,67 @@ const MAX_RESULTS = 10;
+  const [infoVocabObtained, setInfoVocabObtained] = useState("");
+  const [infoVocabTobeObtained, setInfoVocabTobeObtained] = useState("");
+  const [buttonRefs, setButtonRefs] = useState({});
+  const [noResults, setNoResults] = useState(false);
```
```suggestion
```
Unused

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -50,67 +51,67 @@ const MAX_RESULTS = 10;
+function VocabularyQuery(props) {
+  const [editingFilters, setEditingFilters] = useState([]);
+
+  const { classes, defaultValue, disabled, inputRef, noMargin, isNested, onChange, onInputFocus, placeholder, label, value } = props;
```
```suggestion
  const { classes, defaultValue, disabled, noMargin, isNested, onChange, onInputFocus, placeholder, label, value } = props;
```
Unused

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -516,91 +306,180 @@ class VocabularyQuery extends React.Component {
-  clickAwayInfo = (event) => {
-    if ((this.menuPopperRef && this.menuPopperRef.contains(event.target))
-      || (this.infoRef && this.infoRef.contains(event.target))) {
+  let clickAwayInfo = (event) => {
```
Unused function now that it's been refactored.

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/infoBox.jsx:
```diff
@@ -0,0 +1,140 @@
+// Component that renders a dialog with term info for a single vocabulary term.
+//
+function InfoBox(props) {
+  const { classes } = props;
```
I recommend unrolling a lot of the props from InfoBox into this line, so that it isn't necessary to prepend `props.` to every variable. (It also makes it more explicit what props are being used)

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -50,67 +51,67 @@ const MAX_RESULTS = 10;
 //  overrideText: When not undefined, this will overwrite the contents of the search bar
 //  defaultValue: Default chosen term ID, which will be converted to the real ID when the vocabulary loads
 //  noMargin: Removes the margin from the search wrapper
```
A few more docstrings should be added:
- isNested
- placeholder (should also be unrolled into the props line)
- value 
- focusAfterSelecting (should also be unrolled into the props line)
- questionDefinition (should also be unrolled into the props line) 
And vocabularyFilter and vocabularies should be removed.

OrdiNeu reviewed (changes requested):
Looks good, just a few suggestions: 
Recommend renaming `query.jsx` to `VocabularyQuery.jsx` to match the naming scheme of the rest of our components.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Addressed codereview comments

sashaandjic reviewed (approved):
tested

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+//  questionDefinition: question definition
+
+function VocabularyQuery(props) {
+  const [editingFilters, setEditingFilters] = useState([]);
```
```suggestion
```
Unused

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+//  overrideText: When not undefined, this will overwrite the contents of the search bar
+//  defaultValue: Default chosen term ID, which will be converted to the real ID when the vocabulary loads
+//  noMargin: Removes the margin from the search wrapper
+//  isNested:  is nested
```
```suggestion
//  isNested:  If true, restyles the element to remove most padding and apply a negative margin for better nesting
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+import PropTypes from "prop-types";
+
+import { withStyles, FormControl } from "@material-ui/core";
+import { Avatar, Button, Card, CardActions, CardContent, CardHeader, ClickAwayListener, Grow, IconButton, Input, InputAdornment, InputLabel } from "@material-ui/core"
```
```suggestion
import { ClickAwayListener, Grow, IconButton, Input, InputAdornment, InputLabel } from "@material-ui/core"
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+
+import { withStyles, FormControl } from "@material-ui/core";
+import { Avatar, Button, Card, CardActions, CardContent, CardHeader, ClickAwayListener, Grow, IconButton, Input, InputAdornment, InputLabel } from "@material-ui/core"
+import { LinearProgress, Link, MenuItem, MenuList, Paper, Popper, Snackbar, SnackbarContent, Tooltip, Typography } from "@material-ui/core";
```
```suggestion
import { LinearProgress, MenuItem, MenuList, Paper, Popper, Snackbar, SnackbarContent } from "@material-ui/core";
```

OrdiNeu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+import { withStyles, FormControl } from "@material-ui/core";
+import { Avatar, Button, Card, CardActions, CardContent, CardHeader, ClickAwayListener, Grow, IconButton, Input, InputAdornment, InputLabel } from "@material-ui/core"
+import { LinearProgress, Link, MenuItem, MenuList, Paper, Popper, Snackbar, SnackbarContent, Tooltip, Typography } from "@material-ui/core";
+import CloseIcon from '@material-ui/icons/Close';
```
```suggestion
```

OrdiNeu reviewed (dismissed):
Looks good

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+//  defaultValue: Default chosen term ID, which will be converted to the real ID when the vocabulary loads
+//  noMargin: Removes the margin from the search wrapper
+//  isNested:  is nested
+//  placeholder: input placeholder
+//  value: input value
+//  questionDefinition: question definition
```
```suggestion
//  placeholder: String to display as the input element's placeholder
//  value: String to use as the input element value
//  questionDefinition: Object describing the Vocabulary Question for which this suggested input is displayed
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+function VocabularyQuery(props) {
+  const [editingFilters, setEditingFilters] = useState([]);
+
+  const { classes, defaultValue, disabled, noMargin, isNested, onChange, onInputFocus,
+    focusAfterSelecting, placeholder, label, value } = props;
```
Add `questionDefinition`, `onClick`, `clearOnClick`, `overrideText`, `placeholder`, and use them without the `props.` prefix.

veronikaslc replied:
I see `placeholder` is already there, did you mean something else?

marta- replied:
No, sorry, I didn't see it.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,498 @@
+  overrideText: '',
+  clearOnClick: true,
+  focusAfterSelecting: true
+};
```
The documentation at the top and these specs don't match.

Must:
* list all **required** props in `propTypes`, and mark them as required
* mention all defaults specified in `defaultProps` in the documentation

Ideal:
* list all props (required and optional) in propTypes

Commit:
LFS-1085: Refactor query.jsx as a functional component
Addressed codereview comments

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,503 @@
+    onClick: PropTypes.func.isRequired,
+    focusAfterSelecting: PropTypes.bool.isRequired,
+    defaultValue: PropTypes.string,
+    disabled: PropTypes.string,
```
This should be `bool`. The documentation mentions it as `Boolean`, and [`MultipleChoice` passes a boolean to it](https://github.com/ccmbioinfo/lfs/blob/dev/modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx#L101).

```suggestion
    disabled: PropTypes.bool,
```

veronikaslc replied:
right

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Corrected `disabled` proptype

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,503 @@
+// Component that renders a search bar for vocabulary terms.
+//
+// Required arguments:
+//  clearOnClick: Whether selecting an option will clear the search bar
```
```suggestion
//  clearOnClick: Whether selecting an option will clear the search bar (default: true)
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,503 @@
+//  clearOnClick: Whether selecting an option will clear the search bar
+//  onClick: Callback when the user clicks on this element
+//  onInputFocus: Callback when the input is focused on
+//  focusAfterSelecting: focus after selecting
```
```suggestion
//  focusAfterSelecting: focus after selecting (default: true)
```

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Added default details to documentation

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Added default details to documentation

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

sashaandjic reviewed (approved):
retested - all good

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,503 @@
+          vocabulary={{url: infoVocabURL, description: infoVocabDescription, acronym: infoVocabAcronym}}
+          closeInfo={closeInfo}
+          term={{name: infoName, id: infoID, definition: infoDefinition, alsoKnownAs: infoAlsoKnownAs}}
+          infoTypeOf={infoTypeOf}
```
Shouldn't this be inside the `term` object?

Commit:
LFS-1085: Refactor query.jsx as a functional component
Group term info together

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,502 @@
+        </Popper>
+        {/* Info box using Popper */}
+        <InfoBox
+          termInfoVisible={termInfoVisible}
```
Rename:
```suggestion
         open={termInfoVisible}
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,502 @@
+          infoRef={infoRef}
+          menuPopperRef={menuPopperRef}
+          vocabulary={{url: infoVocabURL, description: infoVocabDescription, acronym: infoVocabAcronym}}
+          closeInfo={closeInfo}
```
Rename:
```suggestion
          onClose={closeInfo}
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,502 @@
+    }
+  }
+
+  // Grab information about the given ID and populate the info box
+  let getInfo = (path) => {
+    // If we don't yet know anything about our vocabulary, fill it in
+    var vocabPath = `${path.split("/").slice(0, -1).join("/")}.json`;
+    if (infoVocabObtained != vocabPath) {
+      var url = new URL(vocabPath, window.location.origin);
+      setInfoVocabTobeObtained(vocabPath);
+      MakeRequest(url, parseVocabInfo);
+    }
```
Either I'm misunderstanding something, or the comment doesn't match the functionality

veronikaslc replied:
The comment relates to the lines 285 and 286, where the request to the term path returns information about the term and populates the info box via passing the data to the `showInfo` callback. As for the infobox we need vocab info, the additional call is made to obtain it in case it was not fetched before.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,502 @@
+  // callback for getInfo to populate info box
+  let showInfo = (status, data) => {
+    if (status === null) {
+      var typeOf = [];
+      if ("parents" in data) {
+        typeOf = data["parents"].map(element =>
+          element["label"] || element["name"] || element["identifier"] || element["id"]
+        ).filter(i => i);
+      }
+
+      setInfoID(data["identifier"]);
+      setInfoPath(data["@path"]);
+      setInfoName(data["label"]);
+      setInfoDefinition(data["def"] || data["description"] || data["definition"]);
+      setInfoAlsoKnownAs(data["synonyms"] || data["has_exact_synonym"] || []);
+      setInfoTypeOf(typeOf);
+      setInfoAnchor(buttonRefs[data["identifier"]]);
```
We should just pass `data` to the infobox component and do all the parsing there.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,502 @@
+    setTermInfoVisible(false);
+  };
+
+  let openDialog = () => {
+    setBrowserOpened(true);
+    setBrowseID(infoID);
+    setBrowsePath(infoPath);
+  }
+
+  let closeDialog = () => {
```
Rename:
```suggestion
  let openBrowser = () => {
    setBrowserOpened(true);
    setBrowseID(infoID);
    setBrowsePath(infoPath);
  }

  let closeBrowser = () => {
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,502 @@
+  const [infoAnchor, setInfoAnchor] = useState(null);
+
+  // Information about the vocabulary
+  const [infoVocabAcronym, setInfoVocabAcronym] = useState("");
+  const [infoVocabURL, setInfoVocabURL] = useState("");
+  const [infoVocabDescription, setInfoVocabDescription] = useState("");
```
These don't need to have the `info` prefix

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,502 @@
+          browserOpened={browserOpened}
+        />
+        { /* Browse dialog box */}
+        <VocabularyBrowser
+          open={browserOpened}
+          id={browseID}
+          path={browsePath}
+          changeTerm={changeBrowseTerm}
+          onClose={closeDialog}
+          onError={logError}
+          registerInfo={registerInfoButton}
+          getInfo={getInfo}
+          />
```
The only way to the `VocabularyBrowser` from here is through `infoBox`. This should be moved there.

veronikaslc replied:
Some big refactoring is coming.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Extracted term infoBox and vocan browser to a new component for reusability

marta- commented:
Bug discovered while testing #626 for 1115:  patient info form > race/ethnicity: search for a term, click on (i) to open the info box, click on "Learn more" to open the browser => info box stays above vocabulary browser.  Was introduced [here](https://github.com/ccmbioinfo/lfs/pull/627/commits/86e39703aee15544861a987de697278ab87954e8) and addressed requested changes from code review.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Tune clickaway to leave suggestions dropdown after infobox is closed

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBrowser.jsx:
```diff
@@ -0,0 +1,205 @@
+
+// Component that renders a vocabulary info box and browser.
+//
+function InfoBrowser(props) {
```
`InfoBrowser` is not an informative enough name. After spending HOURS trying to find a bettter one, I suggest to:
* rename `InfoBrowser` to `VocabularyBrowser`, since it contains the actual browsing functionality
* rename the current VocabularyBrowser as `VocabularyTree` since it basically displays the structure
* rename the current `ListChild` from `browseListChild.jsx` to `VocabularyBranch`
* also rename the files to match the component names

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBrowser.jsx:
```diff
@@ -0,0 +1,205 @@
+        { browseID && browserOpened && <VocabularyBrowser
+          browserRef={browserRef}
+          open={browserOpened || false}
+          id={browseID}
```
Please correct me if I'm wrong, but after reading the code all the way down to `browseListChild.jsx` I can't find a good reason to pass this `id` around. The `path` is also unique for each term and we can identify the term by it, reducing the redundancy and the number of props.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBrowser.jsx:
```diff
@@ -0,0 +1,205 @@
+
+  // Strings used by the info box
+  const [infoAboveBackground, setInfoAboveBackground] = useState(false);
+  const [infoAnchor, setInfoAnchor] = useState(null);
```
These are not strings. `Info box properties`?
```suggestion
  // Info box properties
  const [infoAboveBackground, setInfoAboveBackground] = useState(false);
  const [infoAnchor, setInfoAnchor] = useState(null);
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBrowser.jsx:
```diff
@@ -0,0 +1,205 @@
+      if ("parents" in data) {
+        typeOf = data["parents"].map(element =>
+          element["label"] || element["name"] || element["identifier"] || element["id"]
+        ).filter(i => i);
```
What is the purpose of this `filter`? It doesn't seem necessary since `map` already makes copies.

Can this whole block actually be replaced with:

```suggestion
      data["parents"]?.map(p => p["label"] || p["name"] || p["identifier"] || p["id"]) || []
```

and moved down to line 119, for consistency with the other term entries?

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBrowser.jsx:
```diff
@@ -0,0 +1,205 @@
+    }
+  }
+
+  let changeBrowseTerm = (id, path) => {
```
If it turns out we can indeed remove `browseID` per previous comment, this function becomes unnecessary and `setBrowsePath` could be used directly instead.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBrowser.jsx:
```diff
@@ -0,0 +1,205 @@
+          open={browserOpened || false}
+          id={browseID}
+          path={browsePath}
+          changeTerm={changeBrowseTerm}
```
Rename the `changeTerm` prop to `onTermFocus`

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBrowser.jsx:
```diff
@@ -0,0 +1,205 @@
+import { withStyles } from "@material-ui/core";
+
+import VocabularyBrowser from "./browse.jsx";
+import InfoBox from "./infoBox.jsx";
```
Please capitalize file names to match the name of the component they define.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,384 @@
+        setSuggestionsVisible(true);
+        setSuggestionsLoading(false);
+    } else {
+      console.log("Failed to search vocabulary term");
```
Regression: since the `SnackBar` was completely moved to `InfoBox` we no longer display to the user any message when the suggestions cannot be loaded, and they are left looking at the "loading" input with zero feedback.

I suggest an `error` state on the input and a `helperText` that reads "Cannot load suggestions. Please inform your administrator." (we may adjust the text in another iteration).

marta- reviewed (changes requested):
Looking good, just a few more change suggestions.

marta- commented:
Bug:
* In Race/Ethnicity, type something to get suggestions
* click on (i) for one of the suggestions
* click on "Learn more" to bring up the browser
* click on (i) for one of the items in the vocabulary tree to display the info box, then click on (x) to close the infobox
* click on (i) for another one of the items the info box, then click on (x) to try to close it -> **nothing happens**

marta- commented:
Bug / usability issue: opening the vocabulary browser hides the suggestions and clears the input.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Addressed codereview comments

veronikaslc commented:
@marta- Ready for re-review

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/BrowseVocabularyBranch.jsx:
```diff
@@ -216,10 +216,10 @@ function ListChild(props) {
 };
 
-const BrowseListChild = withStyles(BrowseTheme)(ListChild);
+const BrowseVocabularyBranch = withStyles(BrowseTheme)(VocabularyBranch);
```
Don't rename for export
```suggestion
export default withStyles(BrowseTheme)(VocabularyBranch);
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/BrowseVocabularyBranch.jsx:
```diff
@@ -216,10 +216,10 @@ function ListChild(props) {
 
-export default BrowseListChild;
\ No newline at end of file
+export default BrowseVocabularyBranch;
```
```suggestion
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/BrowseVocabularyBranch.jsx:
```diff
@@ -46,8 +46,8 @@ import { REST_URL, MakeRequest } from "./util.jsx";
-function ListChild(props) {
-  const { classes, defaultOpen, id, path, name, changeTerm, registerInfo, getInfo, expands, headNode, bolded, onError, knownHasChildren } = props;
+function VocabularyBranch(props) {
+  const { classes, defaultOpen, id, path, name, onTermFocus, registerInfo, getInfo, expands, headNode, bolded, onError, knownHasChildren } = props;
```
Rename `bolded` to `focused`

```suggestion
  const { classes, defaultOpen, id, path, name, onTermFocus, registerInfo, getInfo, expands, headNode, focused, onError, knownHasChildren } = props;
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/infoBox.jsx:
```diff
@@ -0,0 +1,129 @@
+//
```
This file hasn't been renamed to `InfoBox.jsx`.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Better component naming

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -297,6 +298,7 @@ function VocabularyQuery(props) {
               { (document.activeElement === anchorEl.current || (!defaultValue && !(anchorEl.current?.value))) ? label : ''}
             </InputLabel>
             {inputEl}
+            { error && <FormHelperText>{error}</FormHelperText> }
```
Doesn't look great:
![image](https://user-images.githubusercontent.com/651980/118269926-4b33b380-b48d-11eb-9ffa-2c87d7d47339.png)

* The progress bar should disappear once we know we can't load the suggestions
* The helperText should also be red
* The input should retain its original size that is consistent with the size of all other inputs in the form, instead of stretching to the length of the helper text.

marta- replied:
> * The input should retain its original size that is consistent with the size of all other inputs in the form, instead of stretching to the length of the helper text.

This part is not resolved in the last commit.

marta- replied:
Resolved in c843b86 .

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,129 @@
+                    </div>
+                  )}
+                  </CardContent>
+                  <CardActions className={classes.infoPaper + (browserOpened ? " " + classes.hiddenButton : '')}>
+                      <Button size="small" onClick={openBrowser} variant='contained' color='primary'>Learn more</Button>
+                  </CardActions>
```
I never liked the fact that the "Learn more" button disappeared in the background when the browser was open, and now we have an opportunity to improve it. We're not supposed to show "Learn more" when displaying the info for a term in the Vocabulary Browser, but it should not affect what the info box for a VocabularyQuery suggestion looks like.

Rather than hide the button when the browser opens, we should only display it if info box is not above background.

Would this work?
```suggestion
                  <CardActions className={classes.infoPaper}>
                    { !(broswerOpened && infoAboveBackground) &&
                      <Button size="small" onClick={openBrowser} variant='contained' color='primary'>Learn more</Button>
                    }
                  </CardActions>
```

marta- replied:
Addressed with changes (hid the whole DialogActions, fixed typo in var name) in 309a2e29.

marta- reviewed (changes requested):
Bug:
* type some text in the input to trigger suggestions
* click in (i) for one of the suggestions
* click on "Learn more"
* click on one of the parents
Result: **white screen and JS console error**
```
Uncaught TypeError: Cannot read property 'branch' of undefined
    at VocabularyBranch (VocabularyBranch.jsx:241)
   ...
```

EDIT: **resolved**

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,386 @@
+        setSuggestionsVisible(true);
+        setSuggestionsLoading(false);
+    } else {
+      setError("Cannot load suggestions. Please inform your administrator.");
```
```suggestion
      setError("Cannot load answer suggestions for this question. Please inform your administrator.");
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -193,7 +193,7 @@ function ListChild(props) {
       {/* Term name */}
       <Typography onClick={() => loadTerm(id, path)}
-                  className={classes.infoName + (bolded ? (" " + classes.boldedName) : " ")}>
+                  className={classes.infoName + (focused ? (" " + classes.boldedName) : " ")}>
```
Rename the class name:
```suggestion
                  className={classes.infoName + (focused ? (" " + classes.focusedTermName) : " ")}>
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -30,24 +30,24 @@ import More from "@material-ui/icons/MoreHoriz";
 //  expands: boolean determining whether or not to allow this child to display its children
 //  headNode: boolean determining whether or not this node is the topmost node in the browser
-//  bolded: boolean determining whether or not to bold this entry
+//  focused: boolean determining whether or not to bold this entry
```
```suggestion
//  focused: boolean determining whether this entry is focused and should be visually emphasized
//           (a focused term entry is displayed as a root of a subtree, with only its parents above and its descendants below)
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBrowser.jsx:
```diff
@@ -0,0 +1,191 @@
+import { REST_URL, MakeRequest } from "./util.jsx";
+import QueryStyle from "./queryStyle.jsx";
+
+// Component that renders a vocabulary info box and browser.
+//
```
Please add documentation here and `propTypes` below.

marta- replied:
Same for all other components touched by this refactoring.

veronikaslc replied:
I propose first to settle on the refactoring structure and used params and make documentation a very last step of the PR.

marta- replied:
Fair enough, but I often realize the params need refactoring by looking at the documentation so it's not guaranteed it will be the last step as it may trigger additional change requests.

marta- commented:
> Bug:
> 
> * type some text in the input to trigger suggestions
> * click in (i) for one of the suggestions
> * click on "Learn more"
> * click on one of the parents
>   Result: **white screen and JS console error**
> 
> ```
> Uncaught TypeError: Cannot read property 'branch' of undefined
>     at VocabularyBranch (VocabularyBranch.jsx:241)
>    ...
> ```

Fixed in e2a5c0e .

marta- committed:
LFS-1085: Refactor query.jsx as a functional component
Fix regression: white screen when attempting to focus on a different term in the Vocabulary browser
Caused by not not renaming the `VocabularyBranch` component upon export, and calling it recursively without `classes`

marta- committed:
LFS-1085: Refactor query.jsx as a functional component
Improvement: Don't let "Leanrn more" disappear in the background infor box when opening the Vocabulary Browser

marta- commented:
One more regression / UX annoyance: in the browser open an infobox and ten click on another term -> the info box remains open.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,131 @@
+// Component that renders a dialog with term info for a single vocabulary term.
+//
+function InfoBox(props) {
+  const { open, infoboxRef, vocabulary, onClose, term, infoAboveBackground, browserOpened, openBrowser, onClickAway, classes } = props;
```
Rename openBrowser to `onActionClick` (use the `on<Event>` naming convention to clarify that property is a fuction).

Reorder props for better readability.
```suggestion
  const { open, infoboxRef, vocabulary, term, infoAboveBackground, browserOpened, onActionClick, onClickAway, onClose, classes } = props;
```

Commit:
LFS-1085: Refactor query.jsx as a functional component
Addressed codereview questions

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBrowser.jsx:
```diff
@@ -151,7 +156,7 @@ function VocabularyBrowser(props) {
           onClose={closeInfo}
           term={term}
-          openBrowser={openBrowser}
+          onActionClick={onOpenBrowser}
```
Only the prop needs to be changed, not the function name.

Rationale:
* the prop should be named to specify _when_ the handler is called (e.g. `onClose`, `onClick`, `onClickAway`, etc);
* the value of the prop is a function that should be named to specify _what_ it does (e.g. `onClose: clearData`, `onClick: openDialog`).

marta- replied:
Resolved in fe982d3 .

marta- committed:
LFS-1085: Refactor query.jsx as a functional component
Minor renaming and reordering for better readability

marta- committed:
LFS-1085: Refactor query.jsx as a functional component
Fixed the size of suggestion input when the error message is displayed.

marta- reviewed (commented):
All functionality and most code style seems good now. Needs documentation.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/queryStyle.jsx:
```diff
@@ -192,7 +192,7 @@ const thesaurusStyle = theme => ({
   infoDialog: {
     zIndex: "1350 !important",
-  }
+  },
```
Artifact change

sashaandjic commented:
- [ ] feedback to the user when vocabulary is not yet installed and the user is typing in the suggested input
- [ ] adding a non-predefined answer to vocabulary questions with input
- [ ] getting the information about a term by clicking the info button
- [ ] browsing a vocabulary
- [ ] filtering by vocabulary questions in the Forms table

![image](https://user-images.githubusercontent.com/23243244/118380127-82998180-b5ad-11eb-8547-b4e5988cf621.png)
![image](https://user-images.githubusercontent.com/23243244/118380274-76fa8a80-b5ae-11eb-8146-cbfa99cc1adc.png)
![image](https://user-images.githubusercontent.com/23243244/118380318-cb9e0580-b5ae-11eb-99db-aad7afbd6522.png)
![image](https://user-images.githubusercontent.com/23243244/118380339-03a54880-b5af-11eb-9500-b59a2df67490.png)

sashaandjic commented:
Why is `Croatian `not showing as:
`European `
--> `Croatian `
--> --> `Split founder` 
--> --> `Vis founder `

Split founder and Vis founder are shown under European, but not Croatian.
![image](https://user-images.githubusercontent.com/23243244/118380502-33a11b80-b5b0-11eb-80b2-37fee9edfc2e.png)
![image](https://user-images.githubusercontent.com/23243244/118380505-3ac82980-b5b0-11eb-9ab2-a9d01bbd1cb2.png)
![image](https://user-images.githubusercontent.com/23243244/118380510-40257400-b5b0-11eb-89a6-5f5cb6c97a97.png

marta- commented:
> Why is `Croatian` not showing as:
> `European`
> --> `Croatian`
> --> --> `Split founder`
> --> --> `Vis founder `
> 
> **As per info box, Split founder and Vis founder should be shown first under Croatian, then European.**

`Croatian` is not defined in the ontology as a descendant of `European`, it appears directly under `undefined ancestry population`. The code displays the structure defined by this vocabulary correctly, if anything there's an issue with the vocabulary itself and how populations are organized.

sashaandjic reviewed (approved):
tested

marta- commented:
Noticed a very elusive bug:
* type some text in a suggested input, let the suggestions pop up
* let it sit for a few minutes
* then click on (i) a couple of times

Result: at some point the info box ends up stuck to the top left corner of the page and there's a JS console warning complaining `anchorEl` is invalid:

```
Warning: Failed prop type: Material-UI: The `anchorEl` prop provided to the component is invalid.
The anchor element should be part of the document layout.
```
![image](https://user-images.githubusercontent.com/651980/118399865-2b3af600-b62d-11eb-92a7-7de0d6d3d320.png)

Commit:
LFS-1085: Refactor query.jsx as a functional component
fixed artifact edit; Return error on logout for term search

Commit:
LFS-1085: Refactor query.jsx as a functional component
Added documentation

veronikaslc commented:
> Noticed a very elusive bug:
> 
> * type some text in a suggested input, let the suggestions pop up
> * let it sit for a few minutes
> * then click on (i) a couple of times
> 
> Result: at some point the info box ends up stuck to the top left corner of the page and there's a JS console warning complaining `anchorEl` is invalid:
> 
> ```
> Warning: Failed prop type: Material-UI: The `anchorEl` prop provided to the component is invalid.
> The anchor element should be part of the document layout.
> ```
> 
> ![image](https://user-images.githubusercontent.com/651980/118399865-2b3af600-b62d-11eb-92a7-7de0d6d3d320.png)

Could not reproduce the issue in any timeouts. Added a better logout handling just in case this is the cause.

marta- committed:
LFS-1085: Refactor query.jsx as a functional component
Renamed prop `onTermFocus` to `onTermClick`

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,156 @@
+// open: Boolean representing whether or not the info dialog is open
+// infoboxRef: Reference to the info dialog node
+// vocabulary: Vocabulary source info
+// term: Vocabulary term of interest
```
```suggestion
// term: Vocabulary term being displayed in the info box
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,156 @@
+//
+// Required arguments:
+// open: Boolean representing whether or not the info dialog is open
+// infoboxRef: Reference to the info dialog node
```
Misleading, the component doesn't display the information in a dialog
```suggestion
// infoboxRef: Reference to the info box node
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,156 @@
+// infoboxRef: Reference to the info dialog node
+// vocabulary: Vocabulary source info
+// term: Vocabulary term of interest
+// infoAboveBackground: Boolean representing whether or not the term info dialog is placed above the vocabulary tree dialog
```
```suggestion
// infoAboveBackground: Boolean representing whether or not the term info box is placed above the vocabulary tree dialog
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,156 @@
+// term: Vocabulary term of interest
+// infoAboveBackground: Boolean representing whether or not the term info dialog is placed above the vocabulary tree dialog
+// browserOpened: Boolean representing whether or not the vocabulary tree dialog is open
+// onActionClick: Callback for the dialog action button click event
```
```suggestion
// onActionClick: Callback for the  action button click event
```

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,156 @@
+// browserOpened: Boolean representing whether or not the vocabulary tree dialog is open
+// onActionClick: Callback for the dialog action button click event
+// onClickAway: Callback for the click away event
+// onClose: Callback for the dialog close event
```
```suggestion
// onClose: Callback for the close event
```

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Renamed prop `onTermFocus` to `onTermClick`

marta- commented:
> Could not reproduce the issue in any timeouts. Added a better logout handling just in case this is the cause.

Just realized it had nothing to do with timeouts but with window resize (in my case when I opened the devtools, but same thing happens on a "normal" resize). On resize the refs are lost.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,389 @@
+    isNested: PropTypes.bool,
+    placeholder: PropTypes.string,
+    value: PropTypes.string,
+    questionDefinition: PropTypes.object,
```
Shouldn't this be required?

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -18,31 +18,33 @@
-//  fullscreen: whether or not the dialog is fullscreen (default: false)
-function VocabularyBrowser(props) {
-  const { classes, fullscreen, id, path, changeTerm, registerInfo, getInfo, onClose, onError, ...rest } = props;
+//  fullscreen: Boolean representing whether or not the dialog is fullscreen (default: false)
```
The documentation says the default is `false`, but the `defaultProps` specify it as `true`.

marta- replied:
Moreover, unless we pass the `fullScreen` setting down from the `VocabularyBrowser` component, it won't actually get changed in practice. I'd just remove it.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Addressed codereview comments

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Removed fullscreen property

marta- reviewed (approved):
Known window resize issue reported separately as https://phenotips.atlassian.net/browse/LFS-1169

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,389 @@
+//  defaultValue: Default chosen term ID, which will be converted to the real ID when the vocabulary loads
+//  noMargin: Removes the margin from the search wrapper
+//  isNested: If true, restyles the element to remove most padding and apply a negative margin for better nesting
+//  placeholder: String to display as the input element's placeholder
```
Just by reading the documentation, I find all these terms confusing: `label`, `overrideText`, `placeholder`, and also `defaultValue` and `value`. Are they all needed? Can the documentation be made more clear as to what each of them does, when it's displayed, and what overrides what?

sdumitriu replied:
`label` and `placeholder` seem to be mutually exclusive. When used in a question, `label` is used. When used as a filter, `placeholder` is used. Could we change the code so that we always use the same prop, `label`, but use it as an `InputLabel` when `noMargin` is false, and as a `placeholder` prop when `noMargin` is true? Is that a good idea @marta- ?

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,389 @@
+//  disabled: Boolean representing whether or not this element is disabled
+//  onInputFocus: Callback when the input is focused on
+//  label: Default text to display in search bar when nothing has been entered (default: 'Search')
+//  overrideText: When not undefined, this will overwrite the contents of the search bar
```
```suggestion
//  overrideText: When not undefined, this will overwrite the contents of the search bar whenever the component is disabled. Useful, for example, to display a message like "Please select at most X options"
```

This would be more accurate. However, this prop is never set on the component, so it could be removed completely. We have a different mechanism for displaying the "select this many answers" message now.

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,389 @@
+//  label: Default text to display in search bar when nothing has been entered (default: 'Search')
+//  overrideText: When not undefined, this will overwrite the contents of the search bar
+//  defaultValue: Default chosen term ID, which will be converted to the real ID when the vocabulary loads
+//  noMargin: Removes the margin from the search wrapper
```
Bad name, since more than just the margin is affected. How about `variant: standard|inline`, defaulting on `standard`?

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,389 @@
+          event.preventDefault();
+        }
+      }}
+      onFocus={(status) => {
```
Regression: install HANCESTRO, dashboard, add filter, race/ethnicity, select an item, close, edit filters, focus on the input -> value disappears. The buggy code doesn't seem to be here, though.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/InfoBox.jsx:
```diff
@@ -0,0 +1,156 @@
+                subheader={term.id}
+                titleTypographyProps={{variant: 'h5'}}
+              />
+              <CardContent className={classes.infoPaper}>
```
Awkward display of info for terms with long names, because  `max-width: 500px` is specified by `infoPaper` on the `CardContent` rather than the `Card` itself:
![image](https://user-images.githubusercontent.com/651980/119063660-a68f0580-b9a7-11eb-91d5-84cf11556973.png)

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBrowser.jsx:
```diff
@@ -0,0 +1,216 @@
+  }, [infoPath])
+
+  // Event handler for clicking away from the info box or browser while they are open
+  let clickAwayInfo = (event) => {
+    if ( browserRef?.current?.contains(event.target)
+         || infoboxRef?.current?.contains(event.target)) {
+      return;
+    }
+
+    setTermInfoVisible(false);
+    setInfoAboveBackground(false);
+    onClose && onClose();
```
Both @sashaandjic and I think clicking outside of the infoBox, including anywhere on the browser when the infobox is on top, should hide the infoBox.

Would this work?
```suggestion
  let clickAwayInfo = (event) => {
    if ( !infoAboveBackground && browserRef?.current?.contains(event.target)
         || infoboxRef?.current?.contains(event.target)) {
      return;
    }

    setTermInfoVisible(false);
    setInfoAboveBackground(false);
```

veronikaslc committed:
LFS-1085: Refactor query.jsx as a functional component

Clicking outside of the infoBox, including anywhere on the browser when the infobox is on top, should hide the infoBox.

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

Commit:
LFS-1085: Refactor query.jsx as a functional component
Fixed the info box max width on the card level

Commit:
LFS-1085: Refactor query.jsx as a functional component
Use ResponsiveDialog instead of material-ui Dialog component

sashaandjic reviewed (approved):
Awkward display of info for terms with long names is corrected:
![image](https://user-images.githubusercontent.com/23243244/119089050-9727b080-b9d7-11eb-9112-5e6cb322d726.png)

sdumitriu reviewed (changes requested):
The anchor issue is not just related to resizing the window, it can happen just by looking at different terms.

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in VocabularyBranch (created by WithStyles(VocabularyBranch))
    in WithStyles(VocabularyBranch) react_devtools_backend.js:2560:23

    updateChildrenData webpack://lfs-aggregated-frontend/./src/vocabQuery/VocabularyBranch.jsx?:154
    onload webpack://lfs-aggregated-frontend/./src/vocabQuery/util.jsx?:41
    (Async: EventHandlerNonNull)
    MakeRequest webpack://lfs-aggregated-frontend/./src/vocabQuery/util.jsx?:39
    checkForChildren webpack://lfs-aggregated-frontend/./src/vocabQuery/VocabularyBranch.jsx?:146
    VocabularyBranch webpack://lfs-aggregated-frontend/./src/vocabQuery/VocabularyBranch.jsx?:236
```

```
Warning: Failed prop type: Material-UI: The `anchorEl` prop provided to the component is invalid.
The anchor element should be part of the document layout.
Make sure the element is present in the document or that it's not display none.
    in ForwardRef(Popper) (created by InfoBox)
    in InfoBox (created by WithStyles(InfoBox))
    in WithStyles(InfoBox) (created by VocabularyBrowser)
    in VocabularyBrowser (created by WithStyles(VocabularyBrowser))
    in WithStyles(VocabularyBrowser) (created by VocabularyQuery)
    in div (created by VocabularyQuery)
    in VocabularyQuery (created by WithStyles(VocabularyQuery))
    in WithStyles(VocabularyQuery) (created by MultipleChoice)
    in div (created by MultipleChoice)
    in li (created by ForwardRef(ListItem))
    in ForwardRef(ListItem) (created by WithStyles(ForwardRef(ListItem)))
    in WithStyles(ForwardRef(ListItem)) (created by MultipleChoice)
    in ul (created by ForwardRef(List))
    in ForwardRef(List) (created by WithStyles(ForwardRef(List)))
    in WithStyles(ForwardRef(List)) (created by MultipleChoice)
    in MultipleChoice (created by WithStyles(MultipleChoice))
    in WithStyles(MultipleChoice) (created by VocabularyQuestion)
    in div (created by Question)
    in div (created by ForwardRef(CardContent))
    in ForwardRef(CardContent) (created by WithStyles(ForwardRef(CardContent)))
    in WithStyles(ForwardRef(CardContent)) (created by Question)
    in div (created by ForwardRef(Paper))
    in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
    in WithStyles(ForwardRef(Paper)) (created by ForwardRef(Card))
    in ForwardRef(Card) (created by WithStyles(ForwardRef(Card)))
    in WithStyles(ForwardRef(Card)) (created by Question)
    in Question (created by WithStyles(Question))
    in WithStyles(Question) (created by VocabularyQuestion)
    in VocabularyQuestion (created by WithStyles(VocabularyQuestion))

printWarning webpack://lfs-aggregated-frontend/./node_modules/prop-types/checkPropTypes.js?:20
    checkPropTypes webpack://lfs-aggregated-frontend/./node_modules/prop-types/checkPropTypes.js?:82
    React 2
    InfoBox webpack://lfs-aggregated-frontend/./src/vocabQuery/InfoBox.jsx?:78
```

I think this should reproduce it:

- lfs mode, install HANCESTRO, new Patient Information form
- type `g` in Race
- info for Guamanian
- Learn more
- click outside the vocab browser to close it
- info for Guamanian no longer works
- info for another term works
- info for Guamanian now works, but opens in the corner, warning about anchorEl logged

Another behavior:
- type `g` in Race
- info for Guamanian
- Learn more
- click on its parent to open it in the vocab browser
- click outside the vocab browser to close it
- info for another term now opens in a the corner

To reproduce this, it's important to NOT retrigger search before clicking the info button to trigger the error.

Re-launching the search by focusing the search input clears the state enough to make it work properly again.

And there's a bug with focusing the input. Just like it happens for filters (see inline message below), clicking in the search input clears the input text, but the search still uses the old input text. This is a regression.

sashaandjic commented:
Confirming problems reported by @sdumitriu , except for this one that I am unable to reproduce:
- type g in Race ( in the Patient info form; HANCESTRO installed)
- info for Guamanian
- Learn more
- click on its parent to open it in the vocab browser
- click outside the vocab browser to close it
- info for another term now opens in a the corner

Commit:
LFS-1085: Refactor query.jsx as a functional component
Optimised params

sashaandjic commented:
This problem is still present:
- lfs mode, install HANCESTRO, new Patient Information form
- type g in Race
- info for Guamanian
- Learn more
- click outside the vocab browser to close it
- info for Guamanian no longer works
- info for another term works
- info for Guamanian now works

veronikaslc commented:
@sashaandjic I did not commit anything yet, only had changed parameters names. As I mentioned in the standup - still addressing codereview comments.

veronikaslc commented:
@marta-  I am committing changes is parts. What would be the best way to indicate that PR is ready for test? We need an indicator for that so Sasha won't waste time on testing not ready PRs.

sashaandjic commented:
> @marta- I am committing changes is parts. What would be the best way to indicate that PR is ready for test? We need an indicator for that so Sasha won't waste time on testing not ready PRs.

@veronika just mark PR with the light green `testme!` label

Commit:
LFS-1085: Refactor query.jsx as a functional component
Fixed clearing input on focus;
Fixed info box anchor element bug;

Commit:
LFS-1085: Refactor query.jsx as a functional component
Removed unused styling

sashaandjic commented:
All previously reported problems are resolved, however I am unable to modify the filter selection.
![image](https://user-images.githubusercontent.com/23243244/119711146-2fd18c80-be2d-11eb-9172-ba9a03a8b67c.png)

Steps to reproduce:
- install HANCESTRO
- add "`Greek`" to `Race/Ethnicity reported` field and save the form
- go to the Form and set a filter "`Race/Ethnicity reported`" = "`Greek`"
- click on `APPLY `button --> all good
- click either on the filter or on the filter's `"+"` button
![image](https://user-images.githubusercontent.com/23243244/119711703-d0c04780-be2d-11eb-9ff2-229067b159fe.png)
- try to modify "Greek" to something else

marta- commented:
> All previously reported problems are resolved, however I am unable to modify the filter selection.

From an old conversation with @OrdiNeu : "_VocabularyFilter uses defaultValue but not value. Supplying just defaultValue lets the query handle its own internal state, which is useful because the filters don't do it_"

I guess all those props weren't as redundant as they looked @sdumitriu . If in one usecase the component is controlled and in the other one it isn't, we should continue to use either `value` or `defaultValue` depending on the context, and we should keep both props.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Fixed redundancy leftover from params optimisation

veronikaslc commented:
@marta- I had missed one spot to adjust for the params opimisation.

marta- commented:
Noticed a usability issue I thought we had fixed already:
* in Race/Ethnicity type some chars to pop up suggestions
* click on (i) for a suggestion
* click on Learn more to open the browser
* close the browser
 
Result: as the browser is closed, the suggestions and input value disappear as well. This should not happen, the text typed by the user and the suggestions should remain unchanged.

sdumitriu reviewed (commented):
Still a bit buggy:

New Patient Information form, type something in the Race input, then click on the MagnifyingGlass icon -> text in the input is erased, but the search is retriggered with the old input. This used to happen when just focusing on the input (and I complained about it before), so it seems that the fix was only partial.

This is more obvious in the dashboard filter editing: add a filter by any ethnicity, close filters, edit filters, click on the icon. Just as previously reported, the input is cleared, but the old value is still used for searching, and "applying" the filters will not change the previous filter value.

Back on the patient information form, type `r` in the race input, then click on the `i` button for various results. Whenever switching from the info for a term to the next, the info dialog jumps for a little while to the top left corner then moves into the proper place. It's not serious, but the flicker is a bit unnerving. This doesn't happen when switching the term shown from the browser dialog, only with search results.

Patient Information form, type something in the Race input, `i` for a term, `Learn more`. The search results stay open in the background through most interactions, except for this: navigate up the tree until some of the children have an Expand arrow, then click on one of the Expand arrows. I find it odd that (only) expanding a term should close the search suggestions.

None of these issues are major, is it worth waiting for them to be fixed before merging this PR @marta- ?

marta- commented:
> Still a bit buggy:
> 
> New Patient Information form, type something in the Race input, then click on the MagnifyingGlass icon -> text in the input is erased, but the search is retriggered with the old input. This used to happen when just focusing on the input (and I complained about it before), so it seems that the fix was only partial.
> 
> This is more obvious in the dashboard filter editing: add a filter by any ethnicity, close filters, edit filters, click on the icon. Just as previously reported, the input is cleared, but the old value is still used for searching, and "applying" the filters will not change the previous filter value.

This should be fixed.

> 
> Back on the patient information form, type `r` in the race input, then click on the `i` button for various results. Whenever switching from the info for a term to the next, the info dialog jumps for a little while to the top left corner then moves into the proper place. It's not serious, but the flicker is a bit unnerving. This doesn't happen when switching the term shown from the browser dialog, only with search results.

I can't see this in either Chrome or Firefox. Needs a 3rd pair of eyes in front of a slow computer. @sashaandjic , @acrowthe ?

> 
> Patient Information form, type something in the Race input, `i` for a term, `Learn more`. The search results stay open in the background through most interactions, except for this: navigate up the tree until some of the children have an Expand arrow, then click on one of the Expand arrows. I find it odd that (only) expanding a term should close the search suggestions.

I'd like this fixed, along with the similar issue I mentioned in my previous comment.

sdumitriu commented:
> > Back on the patient information form, type `r` in the race input, then click on the `i` button for various results. Whenever switching from the info for a term to the next, the info dialog jumps for a little while to the top left corner then moves into the proper place. It's not serious, but the flicker is a bit unnerving. This doesn't happen when switching the term shown from the browser dialog, only with search results.
> 
> I can't see this in either Chrome or Firefox. Needs a 3rd pair of eyes in front of a slow computer. @sashaandjic , @acrowthe ?
> 
> Edit: for me in FF it flickers a couple of times (in a quite annoying way).

Some more details:

- in FF it happens after every click, while in Chrome it happens only sometimes (1 out of every 3-4 clicks?). My theory is that it happens so fast that sometimes Chrome just doesn't paint the intermediate frames.
- it jumps to the top of the page, not the top of the window, so if the screen is not tall enough and you scroll down far enough to get to the race question, you will not see it jump to the top, just that it flickers; to reproduce it better, create a new questionnaire with just a vocabulary question at the top.

marta- commented:
Noticed

> > > Back on the patient information form, type `r` in the race input, then click on the `i` button for various results. Whenever switching from the info for a term to the next, the info dialog jumps for a little while to the top left corner then moves into the proper place. It's not serious, but the flicker is a bit unnerving. This doesn't happen when switching the term shown from the browser dialog, only with search results.
> > 
> > 
> > I can't see this in either Chrome or Firefox. Needs a 3rd pair of eyes in front of a slow computer. @sashaandjic , @acrowthe ?
> > Edit: for me in FF it flickers a couple of times (in a quite annoying way).
> 
> Some more details:
> 
> * in FF it happens after every click, while in Chrome it happens only sometimes (1 out of every 3-4 clicks?). My theory is that it happens so fast that sometimes Chrome just doesn't paint the intermediate frames.
> * it jumps to the top of the page, not the top of the window, so if the screen is not tall enough and you scroll down far enough to get to the race question, you will not see it jump to the top, just that it flickers; to reproduce it better, create a new questionnaire with just a vocabulary question at the top.

I created a new questionnaire with one vocab question. Can confirm that the flickering is actually the info box jumping to a different position in the page (that isn't within the viewport for me in case of Race/Ethnicity). Also occasionally visible in chrome, see screencast. Should be fixed.

![CPT2105271643-681x871](https://user-images.githubusercontent.com/651980/119894258-629d8280-bf0a-11eb-8596-3a022e1a4273.gif)

sdumitriu commented:
Also, I still get the "anchorEl is not valid" error somehow, though I can't reliably reproduce it (I just noticed it logged after playing with it for a while and I opened the console).

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -0,0 +1,376 @@
+        {props.children}
+
+        <div className={variant == "inline" ? "" : classes.searchWrapper}>
+          {variant == "inline" ?
```
I'm not a fan of the `variant = inline|standard` prop, it's not very clear what each does. ~~I propose a boolean property `withLabel` instead (because this is essentially the difference between the 2~~. Thoughts @sdumitriu ?

marta- replied:
Or to open it up to future use cases: `variant = default|labeled` (and obviously `default` is default). Current `inline` becomes `default`, current `standard` becomes `labeled`.

sdumitriu replied:
I Agree with `variant=default | labeled`.

marta- replied:
@veronikaslc this still needs to be addressed

veronikaslc replied:
@marta- I am confused with the comment and renaming. @sdumitriu requested a values: `standard|inline`, defaulting on `standard`. If per your request current `inline` becomes `default`, current `standard` becomes `labeled` (and obviously `default` is default), that contradicts with the defaulting on `standard` because `standard` becomes `labeled` and not `default`.

marta- replied:
That is correct, we're changing the default and the code depending on the value of `variant` will have to be adjusted.

veronikaslc replied:
Addressed in 29218dc.

veronikaslc commented:
> Still a bit buggy:
> 
> New Patient Information form, type something in the Race input, then click on the MagnifyingGlass icon -> text in the input is erased, but the search is retriggered with the old input. This used to happen when just focusing on the input (and I complained about it before), so it seems that the fix was only partial.
> 
> This is more obvious in the dashboard filter editing: add a filter by any ethnicity, close filters, edit filters, click on the icon. Just as previously reported, the input is cleared, but the old value is still used for searching, and "applying" the filters will not change the previous filter value.

There is a separate Jira issue for that LFS-1174.

sashaandjic commented:
> Noticed a usability issue I thought we had fixed already:
> 
> * in Race/Ethnicity type some chars to pop up suggestions
> * click on (i) for a suggestion
> * click on Learn more to open the browser
> * close the browser by clicking on (x).
> 
> Result: as the browser is closed, the suggestions and input value disappear as well. This should not happen, the text typed by the user and the suggestions should remain unchanged.
> 
> Note: this does not appear to happen when closing the browser by clicking outside.

This is also happening in LFS-1115. Perhaps it was introduced there?

sashaandjic commented:
On WIN machines I guess the jumping is not as prominent. I had to create a new questionnaire to see the jumping in Firefox. It is happening for sure.

https://user-images.githubusercontent.com/23243244/119925883-4585a580-bf44-11eb-8b92-1ac565427cff.mp4

veronikaslc commented:
@sashaandjic All bugs in this PR will be in LFS-1115 because LFS-1115 is based on this PR.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Addressed codereview comments

Commit:
LFS-1085: Refactor query.jsx as a functional component
More clear values for the variant param for VocabularyQuery component

sashaandjic commented:
Only this anomaly remains:
- Create` Patient Information` form and add some value in the `Race/Ethnicity reported` field ( e.g. Eskimo),
- Go to Forms,
- Add a filter by any ethnicity ( e.g. "Eskimo")
- Close filters, 
- Edit filters ( "Eskimo" is shown as expected)
- Click on the magnifying glass icon,
- The input is cleared
- Click on `APPLY ` button
- The old value is still used for searching ("Eskimo", not blank)

https://user-images.githubusercontent.com/23243244/120044045-11a98f00-bfdb-11eb-86e3-b95499333cfb.mp4

veronikaslc commented:
There is a separate Jira issue for that LFS-1174.

marta- commented:
> @sashaandjic There is a separate Jira issue for that LFS-1174.

@veronikaslc not quite. Half of the problem here is that clicking on the magnifying glass erases the input, which shouldn't happen, it should relaunch the search with that input value instead. This is a regression compared to `dev`.

The fact that deleting the filter doesn't have the expected effect is indeed captured by 1174, is also present in dev, and is out of scope for this PR.

Commit:
LFS-1085: Refactor query.jsx as a functional component
Fixed regression in the input search behaviour of the magnifying glass icon

veronikaslc commented:
@marta- Fixed the regression of the input search.

Should we fix LFS-1174 as well? We keep reporting it in the PR, I think 4 times already. May be it falls naturally to be fixed here.

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (changes requested):
One last thing to fix:
- lfs mode
- install HP and ORDO
- new Tumor form
- entering terms in Diagnosis code and Morphology code is broken: in edit mode you can't see what's selected
- another quirk: select values for both of these questions, save, close, reedit -> both terms are visible; focus one of the inputs, both of them are cleared; why does an unrelated question get cleared?

sashaandjic commented:
> * why does an unrelated question get cleared?

Bug when maxAnswers=1

Commit:
LFS-1085: Refactor query.jsx as a functional component
Fixed vocabulary input not showing answers on save/reload/select

sashaandjic reviewed (approved):
Tested

- [x] LFS-1081
- [x] LFS-1151
